diff --git a/Python/compile.c b/Python/compile.c
index 51af28b54f..adc1c651c5 100644
--- a/Python/compile.c
+++ b/Python/compile.c
@@ -893,6 +893,8 @@ stack_effect(int opcode, int oparg, int jump)
         case UNARY_NEGATIVE:
         case UNARY_NOT:
         case UNARY_INVERT:
+        case UNARY_ADDR:
+        case UNARY_DEREF:
             return 0;
 
         case SET_ADD:
@@ -3465,6 +3467,10 @@ unaryop(unaryop_ty op)
         return UNARY_POSITIVE;
     case USub:
         return UNARY_NEGATIVE;
+    case UAddr:
+        return UNARY_ADDR;
+    case UDeref:
+        return UNARY_DEREF;
     default:
         PyErr_Format(PyExc_SystemError,
             "unary op %d should not be possible", op);
